### ISSUE TITLE
JWST dataset has empty images

### ISSUE THREAD

**Issue by LTMeyer** [MultimodalUniverse/MultimodalUniverse#107]
There seem to be several issues with the JWST dataset:

- On rusty, `ngdeep-tiny` has no data;
- `ceers-full` has empty images.

**Comment by LTMeyer** [comment]
```python
f = h5py.File("/mnt/ceph/users/polymathic/MultimodalUniverse/jwst/ceers-full-grizli-v7.0_all_96/healpix=714/001-of-001.hdf5")
image = np.array(f["image_array"][0])
np.allclose(image, 0.)
>>> True
```
Indicates that the images are already null in the original hdf5 file.

**Comment by LTMeyer** [comment]
It seems that adding bands with zero values is part of the process (c.f. https://github.com/MultimodalUniverse/MultimodalUniverse/blob/main/scripts/jwst/build_parent_sample.py#L176-L183 and https://github.com/MultimodalUniverse/MultimodalUniverse/blob/main/scripts/jwst/build_parent_sample.py#L198-L200).
Could someone explain me the logic if returning empty bands there (@EiffL or @mhuertascompany)?

Especially, the following lines mean that as soon as the stamp has more than 10 null data or the wrong shape, we output an empty band.
```python
np.max(stamp.data) <= 0
or np.count_nonzero(stamp.data == 0) > 10
or stamp.data.shape[0] != _image_size
or stamp.data.shape[1] != _image_size
```

**Comment by smericks** [comment]
We've stepped through to [this function call](https://github.com/MultimodalUniverse/MultimodalUniverse/blob/864452cb955464169ae889184f7656a5c0a1de04/scripts/jwst/build_parent_sample.py#L429), and investigated the JWST_stamps object that is saved to a pickle file. It appears that cutouts might be missing a central object after this step. See these example images (log-normed) 
![Screen Shot 2024-09-24 at 3 23 16 PM](https://github.com/user-attachments/assets/637cc7ef-b9e0-49e1-8345-be4e7bb4361e)
![Screen Shot 2024-09-24 at 3 23 29 PM](https://github.com/user-attachments/assets/ccbac75d-1eb3-4e1f-987c-aeb05aee43b3)
![Screen Shot 2024-09-24 at 3 23 54 PM](https://github.com/user-attachments/assets/7f0ecad2-8f3a-4701-ae7c-5aa76d494c1f)

We are looking at ceers-full in filter F115w.

@swagnercarena

**Comment by mhuertascompany** [comment]
Thanks for looking into this. These images are not blank which was the problem we encountered before. There are very faint sources in this dataset since I did not apply any SNR cut. Can you try looking at bright objects by selecting galaxies with F115 magnitudes smaller than ~ 25?

**Comment by mhuertascompany** [comment]
All fluxes are in microJansky.
if you are working with ceers-full, you can use the column f150w_flux_aper_2 for example and compute the mag_AB:
mab = 2.5*alog10(f150w_flux_aper_2*10**-6/3631) and select only objects with mab<25

**Comment by smericks** [comment]
Lines in build_parent_sample.py that we (@swagnercarena and I) found that might be causing issues: 

- Lines [127](https://github.com/MultimodalUniverse/MultimodalUniverse/blob/864452cb955464169ae889184f7656a5c0a1de04/scripts/jwst/build_parent_sample.py#L127), [147](https://github.com/MultimodalUniverse/MultimodalUniverse/blob/864452cb955464169ae889184f7656a5c0a1de04/scripts/jwst/build_parent_sample.py#L147) in _cut_stamps_fn(). When we ran the code locally, we modified these lines to make sure the input variable directory_path was actually used. 

- Line [177](https://github.com/MultimodalUniverse/MultimodalUniverse/blob/864452cb955464169ae889184f7656a5c0a1de04/scripts/jwst/build_parent_sample.py#L177) in _cut_stamps_fn(): Agreed with @LTMeyer, wondering about the logic of the if statement here (especially the first two conditions). 

- Line [297](https://github.com/MultimodalUniverse/MultimodalUniverse/blob/864452cb955464169ae889184f7656a5c0a1de04/scripts/jwst/build_parent_sample.py#L297): Doesn't actually get smallest shape here. Uses variable 'f' that is leftover from a previous loop.

- Line [337](https://github.com/MultimodalUniverse/MultimodalUniverse/blob/864452cb955464169ae889184f7656a5c0a1de04/scripts/jwst/build_parent_sample.py#L337): PSF is hardcoded

**Comment by mhuertascompany** [comment]
Thank you for the tests!
- the issue with noisy empty fields (not nan) is most likely due to faint objects. we could apply a flux cut for the dataset in order to avoid this.
- the first 2 conditions in L.177 were to avoid having objects at the edges of field of view which can be partially empty (number of zero values>10), like the last of your examples. But apparently it is not working properly. 
- I am puzzled by the completely blank images full on nans. I will have a look.

**Comment by mhuertascompany** [comment]
I was wondering if there's been any progress with this. Do you need me to check something else?

**Comment by smericks** [comment]
We wanted to do some tests to figure out why some of the images are all zeros, and I agree it seems like maybe a flux cut could be the next thing to try here. Especially since there are clearly some images (top 3 shown) where everything appears to be working OK. I wonder if someone more familiar with the dataset knows what flux cut would make the most sense.

**Comment by mhuertascompany** [comment]
Thanks!
I'd recommend to look at magntiudes smaller than 25ish
for the flux (f150w_flux_aper_2 in ceers) compute the mag_AB:
mab = 2.5alog10(f150w_flux_aper_210**-6/3631) 
and select only objects with mab<25

**Comment by mb010** [comment]
> Thanks! I'd recommend to look at magntiudes smaller than 25ish for the flux (f150w_flux_aper_2 in ceers) compute the mag_AB: mab = 2.5alog10(f150w_flux_aper_210**-6/3631) and select only objects with mab<25

Did this resolve the issue?

**Comment by mhuertascompany** [comment]
I am not sure it has been tested properly. Could not do it in my end. I have just created a PR in which I have added a couple of extra lines of code to filter out faint objects. It would be nice if someone at the Flatiron could test it to check that it produces better images...Thanks!

**Comment by mb010** [comment]
Ok. So I've been digging around in this for a few days. My conclussions are as follows:
There are lots of empty images because there is lots of empty data (three filters of ngdeep data from DAWN JWST Archiv - our science data source):
![image](https://github.com/user-attachments/assets/0ef5c33f-39b4-47aa-a4b3-d080d55a5b0b)
Some filters have more empty data than others for instrumental / survey reasons. I have yet to see evidence of **excess**  zero data beyond what might actually be expected (after a few bug fixes that were causing issues with readability / excessive flagging / not being able to download files). The catalogs are generated on a particular band for each survey, and so no matter if we place magnitude cuts on them or not, that source might not have any substantial emission in another band.

I will open a PR today with all of my changes (for debugging, downloading and general code quality improvements).

Importantly I found that low bands have a different pixel scale - @mhuertascompany has provided me with a fix which I have merged and it seems to work.

My current plan is to prepare the data in its current format with a note in the ReadME explaining why "so many" zero images appear, and we can decide on if we want to address this for a future version of the jwst data (along with adding other surveys, clusters, and engineering changes).
